drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,543 @@
+/*
+ * Copyright (c) 2020 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This driver emulates a EEPROM device in flash.
+ *
+ * The emulation represents the EEPROM in flash as a region that is a direct
+ * map of the eeprom data followed by a region where changes to the eeprom
+ * data is stored. Changes are written as address-data combinations. The size
+ * of such a combination is determined by the flash write block size and
+ * the size of the eeprom (required address space), with a minimum of 4 byte.
+ * The eeprom page needs to be a multiple of the flash page. Multiple eeprom
+ * pages is also so supported and increases the number of writes that can be
```

Comment:
Yes regarding the comment (have changed this).
Regarding how this is done: the EEPROM is represented in flash as an area of size EEPROM and an area for changes. Whenever a write needs to be performed the library checks if it can do a direct write to the first area (and does so if still empty), if not it stores a change in the second area (address + new data). If there is no more space to store changes a new page is taken, this first writes the first area with the latest values and then starts writing changes.
The ratio between the EEPROM page size and EEPROM size determines the amount of changes that can be written before a page is erased. A EEPROM that would be used to mainly store constant data and only limited changes would use a smaller ratio than a EEPROM that would be used to store frequently changing data. A ratio of 2 between EEPROM page size and EEPROM size (and a partition of 2 EEPROM pages) guarantees that each of the EEPROM cells can be written at least 0.5 * 2 * number of page erases (this is only considering the changes). This is a lot more than in case a write too a next page is performed, this would only allow (number of page erases)/(EEPROM size) changes.